### PR TITLE
ChArUco board

### PR DESCRIPTION
**Pull request by JStech**
Added plugin for a ChArUco board. It'd be helpful to support ChArUco boards, since they're better for intrinsic calibration, and so one target could be used for both.

~~Still needs a test.~~

### CODE REVIEW

**Comment by JStech** [comment]
This is now ready for review, @RoboticsYY and anyone else who's interested (@felixvd and @v4hn?)

**Comment by v4hn** [comment]
Sorry I did not run this calibration package until now, but only gave conceptual feedback.

I would expect @RoboticsYY to review and merge this request.

**Comment by JStech** [comment]
I took a stab at making the `HandEyeTargetBase` class more flexible with regards to different parameters for different boards. I'm not a C++ guru, though, so there might be a better, cleaner (more modern?) way to do it. Definitely open to suggestions.

File: moveit_calibration_plugins/handeye_calibration_target/include/moveit/handeye_calibration_target/handeye_target_base.h [hunk 1]
```diff
@@ -53,6 +54,29 @@ namespace moveit_handeye_calibration
...
+    Parameter(std::string name, ParameterType parameter_type) : name_(name), parameter_type_(parameter_type)
+    {
+    }
+
+    Parameter(std::string name, ParameterType parameter_type, std::vector<std::string> enum_values)
+      : name_(name), parameter_type_(parameter_type), enum_values_(enum_values)
+    {
+    }
+  };
+
```
Comment (JStech) [inline moveit_calibration_plugins/handeye_calibration_target/include/moveit/handeye_calibration_target/handeye_target_base.h:79]: @tylerjw, I'd appreciate your thoughts on this design.

**Comment by JStech** [comment]
moveit rviz- - RViz_024

I have [a branch](github.com/JStech/moveit_calibration/tree/charuco-with-rviz) with the GUI updated to work with this branch.

File: moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+  try
+  {
+    // Create target
+    cv::Ptr<cv::aruco::Dictionary> dictionary = cv::aruco::getPredefinedDictionary(dictionary_id_);
+    cv::Ptr<cv::aruco::CharucoBoard> board = cv::aruco::CharucoBoard::create(
+        markers_x_, markers_y_, float(square_size_pixels_), float(marker_size_pixels_), dictionary);
+
+    // Create target image
+    board->draw(image_size, image, margin_size_pixels_);
```
Comment (RoboticsYY) [inline moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp:165]:
```suggestion
    board->draw(image_size, image, margin_size_pixels_, border_size_bits_);
```

File: moveit_calibration_plugins/handeye_calibration_target/include/moveit/handeye_calibration_target/handeye_target_base.h [hunk 3]
```diff
@@ -182,6 +167,119 @@ class HandEyeTargetBase
     return true;
   }
 
+  virtual std::vector<Parameter> getParameters()
```
Comment (RoboticsYY) [inline moveit_calibration_plugins/handeye_calibration_target/include/moveit/handeye_calibration_target/handeye_target_base.h:170]: Though these parameters related functions are obvious to me, I think it's better to add some Doxygen comment blocks to them.

File: moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+geometry_msgs::Vector3 HandEyeCharucoTarget::convertToVectorROSMsg(const cv::Vec3d& input_tvect) const
+{
+  Eigen::Vector3d eigen_tvect;
+  cv::cv2eigen(input_tvect, eigen_tvect);
+  geometry_msgs::Vector3 msg_tvect;
+  tf2::toMsg(eigen_tvect, msg_tvect);
+  return msg_tvect;
+}
```
Comment (RoboticsYY) [inline moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp:274]: I think the implementations of these three functions are better than the original ones, and they are shared between Aruco and Charuo markers. So there is no need to replicate them. It's better to move these functions to the base class.

File: moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+void HandEyeCharucoTarget::drawAxis(cv::InputOutputArray _image, cv::InputArray _cameraMatrix,
```
Comment (RoboticsYY) [inline moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp:276]: It's better to move this function to base class also. I think any CV based target detection plugin needs this.

File: moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp (hunk 2 above)
Comment (JStech) [reply moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp]: Good catch. Thanks.

File: moveit_calibration_plugins/handeye_calibration_target/include/moveit/handeye_calibration_target/handeye_target_base.h (hunk 3 above)
Comment (JStech) [reply moveit_calibration_plugins/handeye_calibration_target/include/moveit/handeye_calibration_target/handeye_target_base.h:232]: :+1:

File: moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp (hunk 4 above)
Comment (JStech) [reply moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp]: Done

File: moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp (hunk 5 above)
Comment (JStech) [reply moveit_calibration_plugins/handeye_calibration_target/src/handeye_target_charuco.cpp]: Done

**Comment by JStech** [comment]
OK, the tests now pass CI. I aslo changed the parameter names a bit to try to clearly communicate what they mean.

I've got nothing more to add at this point.

**Pull request merged**